### PR TITLE
Add deep-merge of values, and allow empty values file

### PR DESCRIPTION
**Pull request by Lingkar**
Allow empty YAML values files.
Perform a deep merge instead of shallow, for better helm values patching.

(I was unable to locally run the tests, would like to check via the pipeline)

### CODE REVIEW

**Comment by allenporter** [comment]
Hello, appreciate the contribution!

I'd like there to be tests reproducing the problem you've run into here. Can you tell me more about why you were unable to run the tests?

**Comment by Lingkar** [comment]
@allenporter Seems like the issue was related to missing system dependency kyverno.
I will add some issue specific tests, will notify you when ready.

**Comment by Lingkar** [comment]
@allenporter I have updated the values tests to include an empty values and a patch values config.
Please let me know if I am missing anything else!

File: flux_local/values.py [hunk 1]
```diff
@@ -173,6 +173,19 @@ def _lookup_value_reference(
 
     return found_value
 
+def deep_merge(base: dict, override: dict) -> dict:
+    """
+    Recursively merge two dictionaries, similar to how Helm merges values.
+    Lists are replaced entirely (Helm behavior).
+    """
```
Comment (allenporter) [inline flux_local/values.py:180]: Style in this code base is:

```suggestion
    """Recursively merge two dictionaries, similar to how Helm merges values.

    Lists are replaced entirely (Helm behavior).
    """
```

File: flux_local/values.py [hunk 2]
```diff
@@ -173,6 +173,19 @@ def _lookup_value_reference(
...
+    result = base.copy()
+    for key, value in override.items():
+        if key in result and isinstance(result[key], dict) and isinstance(value, dict):
```
Comment (allenporter) [inline flux_local/values.py:183]: Extract `result.get(key)` into a variable since it is looked up 3 times:

**Review by allenporter — CHANGES_REQUESTED** [review]
Looks great, though I have some minor nits. Also there are lint errors so you'll want to check those. Running run the pre-commit before committing can help catch those.

If the yaml lint doesn't like this and it can't be tested that way easily, you can also use `test_values.py` since it has some existing tests for this function.

File: flux_local/values.py (hunk 2 above)
Comment (Lingkar) [reply flux_local/values.py]: Extracted into a variable

**Comment by Lingkar** [comment]
@allenporter Thanks for the pre-commit tip! Might be a good idea to update the CONTRIBUTING.md with the system dependency kyverno and the usage of pre-commit?

I have processed the feedback in the PR.
When running pre-commit run --all-files I got some additional files which got updated, I left those out of my commits as unsure if I am supposed to run the pre-commit tool on all-files or just on the changed files in my PR.

**Comment by allenporter** [comment]
> @allenporter Thanks for the pre-commit tip! Might be a good idea to update the CONTRIBUTING.md with the system dependency kyverno and the usage of pre-commit?

As for the pre-commit, yes, good idea, feel free to send a change to update it.
I think I may want to just drop kyverno support so we can leave that out and i'll remove it separatly. Thanks for this suggestion.


> 
> I have processed the feedback in the PR. When running pre-commit run --all-files I got some additional files which got updated, I left those out of my commits as unsure if I am supposed to run the pre-commit tool on all-files or just on the changed files in my PR.

Let's stick with just the files in this PR.

**Comment by allenporter** [comment]
Thanks for the contribution @Lingkar 💯

**Comment by allenporter** [comment]
I've released this as 6.1.0 and it's causing some HelmReleases to break https://github.com/onedr0p/home-ops/actions/runs/12473136322/job/34813129881

**Comment by allenporter** [comment]
It appears like the original `values` are getting dropped https://github.com/onedr0p/home-ops/blob/aa26601e17589c8f4fc88d6a68503d53c9943c25/kubernetes/main/apps/kube-system/cilium/app/helmrelease.yaml#L30

**Pull request merged**